Feat: Auto TPI - Automatic TPI Coefficient Learning & Calibration

KipK opened the pull request:
This major update introduces the "Auto TPI" feature, allowing the thermostat to automatically learn and optimize its TPI coefficients (Indoor/Outdoor) over learning sessions or continuous learning.

Key Changes:
- **Core Logic**: Introduced AutoTpiManager to orchestrate the entire self-learning lifecycle, enabling continuous adaptation, regime change detection, and state persistence for TPI coefficients..
- **Algorithms**: Added support for EMA (Adaptive Alpha) and Weighted Average learning methods to smooth coefficient adjustments.
- **Refactor**: Refactored base_thermostat.py , introducing new "thermostat_tpi" class. All Vtherm using TPI regulation are now Thermostat_TPI instances.
- **Calibration**: Introduce service action "calibrate_capacity" to estimate adiabatic capacity using the 75th percentile of slopes and Kext compensation.
- **Integration**: Deeply integrated with `Thermosat_TPI` class to intercept cycles and apply learned parameters to `PropAlgorithm` dynamically.
- **Configuration**: Enhanced Config Flow with dedicated Auto TPI steps (activation, method selection, manual capacity override).
- **Sensors**: Added `auto_tpi_learning_state` sensor to monitor learning state, confidence, and stability.
- **Documentation**: Added extensive developer documentation (`auto_tpi_internal_doc.md`) and user guides (FR/EN).
- **Tests**: Comprehensive test suite (`test_auto_tpi.py`) covering initialization, learning logic, and persistence.

Thanks to @gael1980 for the support and the scientific endorsement.

**Note:** A **custom lovelace card** is available [here ](https://github.com/KipK/auto-tpi-learning-card)to follow auto-tpi learning sessions and provide consolidated data for debug:

<img width="515" height="493" alt="image" src="https://github.com/user-attachments/assets/3932e25c-1f05-4b03-b11a-9a4c87ac067b" />

[auto-tpi-learning-card](https://github.com/KipK/auto-tpi-learning-card)

KipK commented:
J'essaye de corriger les tests. C'est un peu le bordel. 

Je galère encore avec certains comme les tests du boiler. Mais je suis pas sûr que ce soit suite au refactor.

KipK commented:
Le dernier commit devrait etre ok pour le refactor climate_valve.

374 test passent maintenant. Il en reste 4. 
Dont central_boiler mais je crois qu'il a jamais marché chez moi. Je  te laisse celui là ;)

KipK commented:
Tous les tests sont ok, sauf central_boiler que je te laisse car j'ai pas trouvé ce qui se passe.
Je te laisse vérifier que j'ai pas fait de bêtises. Ca m'a l'air ok.

KipK commented:
Roger, we're finally all green !
<img width="132" height="230" alt="image" src="https://github.com/user-attachments/assets/e6a5a378-933e-49c0-aa76-387f0b76b1c6" />

KipK commented:
Enfin, green. J'ai besoin de toi sur ce coup là, j'ai aucune idée pour l'instant. mais depuis mes derniers changements, j'ai la plupart des entités qui se sont dupliquées/renommées. Ca vient d'arriver. 

<img width="312" height="612" alt="image" src="https://github.com/user-attachments/assets/96e91a65-22f8-47c2-9bda-8cba2b43aa0f" />

jmcollin78 commented:
Dans underlyings.py méthode remove_entity est-ce qu'il ne faut pas supprimer les `_on_cycle_start_callbacks` ? C'est pas vraiment des listeners on est d'accord et normalement l'objet n'est pas recyclé mais ca mange pas de pain de raz les callback. Je pense aux tests surtout

jmcollin78 commented on custom_components/versatile_thermostat/feature_power_manager.py:
```diff
@@ -177,9 +177,13 @@ def calculate_power_consumption_max(self) -> float:
             if self._vtherm.is_over_climate:
                 power_consumption_max = self._device_power
             else:
+                on_percent = 0
```
On doit pouvoir faire un BaseThermostat.safe_on_percent qui renvoie 0 et qui est surchargé dans ThermostatTPI avec la logique des if return 0. Ca va fluidifier ça partout où c'est nécessaire.

jmcollin78 commented on custom_components/versatile_thermostat/sensor.py:
```diff
@@ -118,6 +120,13 @@ async def async_setup_entry(
                 RegulatedTemperatureSensor(hass, unique_id, name, entry.data)
             )
 
+        # Add Auto TPI Sensor for TPI-based thermostats
```
Dans le même genre on doit pouvoir ajouter un has_tpi dans BaseThermostat qui vaut False et qui passe à True dans ThermostatTPI. Je suis assez de ce genre d'attribut qui simplifie vachement la lecture

KipK replied:
Carrément. C'est fait.

jmcollin78 reviewed (commented):
On est quasi bon pour moi.

KipK commented:
Pour info, y a bien 2 commits qui sont pas liés à ce PR, qui reglent des pétouilles 

-[ là](https://github.com/jmcollin78/versatile_thermostat/pull/1396/commits/aec94f6c6409cef5e747551c51311025060040e6)

- et [là ](https://github.com/jmcollin78/versatile_thermostat/pull/1396/changes/c2674b47588d4eb465b4b3f7b0f2971cec083a55)

Tu me diras si tu veux qu'on les sorte du PR. Je t'avoue,, que si on peut s'en passer ça me va ;)

KipK commented on custom_components/versatile_thermostat/sensor.py:
```diff
@@ -583,7 +583,7 @@ async def async_my_climate_changed(self, event: Event = None):
         if self._attr_native_value is None or self._attr_native_value == 0:
             return "mdi:thermometer"
-        elif self._attr_native_value > 0:
+        elif isinstance(self._attr_native_value, (int, float)) and self._attr_native_value > 0:
```
C'est juste pour résoudre une erreur pylint, le dernier rouge qui pique Je te le mets dans le premier cas? ou on garde le rouge. :)

jmcollin78 replied:
Ici il faut resolve.

KipK commented on custom_components/versatile_thermostat/__init__.py:
```diff
@@ -342,6 +342,14 @@ async def async_migrate_entry(hass: HomeAssistant, config_entry: ConfigEntry):
             new[CONF_UNIQUE_ATTR_ID] = config_entry.data.get(CONF_NAME)
 
-        _LOGGER.info("Migration to version %s successful", config_entry.version)
+        # Update the config entry with migrated data
```
Oui les allers retours entre version, ca m'a foutu un petit bordel déjà. ( mais c'était déjà le cas avant que je rajoute ça ). Ca mérite une reflexion plus profonde.

KipK commented:
Petit ajout dans le config flow:

<img width="604" height="610" alt="image" src="https://github.com/user-attachments/assets/bfc7956c-35be-482a-9872-1edada2d07e7" />

jmcollin78 commented:
Please rebase and resolve conflict since I revert a PR.

KipK commented:
I think we're all good now. Rebased, cleaned up, and new calibrate_capacity service call algo works far better. 

Translations keys for other language than en/fr still needs to be done. Perhaps it's better to add them only at the end of beta test in case we have structural changes in the config_flow, Let's not add'up too much work
